index.mjs

```diff
@@ -0,0 +1,203 @@
+  const { logDir, changesDir, runDir } = osmdbtConfig;
+  const backupDir = path.join(changesDir, 'backup');
+  [logDir, changesDir, runDir, backupDir].forEach(async (dir) => await createDirectory(dir));
+};
+
+const streamToString = (stream) => {
+  return new Promise((resolve, reject) => {
+    const chunks = [];
+    stream.on('data', (chunk) => chunks.push(chunk));
+    stream.on('error', reject);
+    stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
+  });
+};
+
+const getSequenceNumber = async () => {
+  const stateFileContent = await fs.readFile(OSMDBT_STATE_PATH, 'utf8');
+  const matchResult = stateFileContent.match(/sequenceNumber=\d+/);
+  if (matchResult === null || matchResult.length === 0) {
+    throw new ErrorWithExitCode(`failed to fetch sequence number out of the state file, ${STATE_FILE} is invalid`, ExitCodes.INVALID_STATE_FILE);
+  }
+  return matchResult[0].split('=')[1];
+};
+
+const getDiffDirPathComponents = (sequenceNumber) => {
+  const top = sequenceNumber / DIFF_TOP_DIR_DIVIDER;
+  const bottom = (sequenceNumber % DIFF_TOP_DIR_DIVIDER) / DIFF_BOTTOM_DIR_DIVIDER;
+  const state = sequenceNumber % DIFF_STATE_FILE_MODULO;
+  return [top, bottom, state].map((component) => {
+    const intComponent = parseInt(component);
+    return intComponent.toString().padStart(3, '0');
+  });
+};
+
+const initializeS3Client = () => {
+  const { endpoint } = objectStorageConfig;
+  return new S3Client({
+    signatureVersion: 'v4',
+    endpoint,
+    region: S3_REGION,
+    forcePathStyle: true,
+  });
+};
+
+const getStateFileFromS3ToFs = async () => {
+  const bucketName = objectStorageConfig.bucketName;
+  try {
+    const stateFileStream = await s3Client.send(new GetObjectCommand({ Bucket: bucketName, Key: STATE_FILE }));
+    const stateFileContent = await streamToString(stateFileStream.Body);
+    await fs.writeFile(OSMDBT_STATE_PATH, stateFileContent);
+    await fs.writeFile(OSMDBT_STATE_BACKUP_PATH, stateFileContent);
+  } catch (error) {
+    throw new ErrorWithExitCode(
+      `failed getting key: ${STATE_FILE} from bucket: ${bucketName} received the following error: ${error}`,
+      ExitCodes.STATE_FETCH_FAILURE_ERROR
+    );
+  }
+};
+
+const runOsmdbtCommand = async (command, commandArgs = [], onError = undefined) => {
+  logger.info(`preparing run of ${command}`);
+  try {
+    const args = [...GLOBAL_OSMDBT_ARGS, ...commandArgs];
+    const commandPath = path.join(OSMDBT_BIN_PATH, command);
+    await $`${commandPath} ${args}`;
+  } catch (error) {
+    if (onError) {
+      await onError();
+    }
+    throw new ErrorWithExitCode(error.stderr, ExitCodes.OSMDBT_ERROR);
+  }
+};
+
+const putObjectWrapper = async (key, body) => {
+  const { bucketName, acl } = objectStorageConfig;
+  try {
+    logger.info(`putting key: ${key} into bucket: ${bucketName}`);
+    await s3Client.send(new PutObjectCommand({ Bucket: bucketName, Key: key, Body: body, ACL: acl }));
```

Comment:
you should add content-type if its possible (seems like it should but not sure how https://github.com/aws/aws-sdk-js-v3/issues/1916)

Reply:
done.